### PR TITLE
More SVE instructions

### PR DESCRIPTION
**Pull request by maranget**
Add the AArch64 SVE instructions `RDVL` and `ADDVL`.

### CODE REVIEW

File: herd/tests/instructions/AArch64.sve/V23.litmus [hunk 1]
```diff
@@ -0,0 +1,10 @@
+AArch64 V23
+{
+ uint64_t 0:X0;
+ uint64_t 0:X1;
+}
+  P0       ;
+RDVL X0,#0 ;
+RDVL X1,#1 ;
+
+forall 0:X0=0 /\ 0:X1=8
```
Comment (relokin) [inline herd/tests/instructions/AArch64.sve/V23.litmus:10]:
```suggestion
forall 0:X0=0 /\ 0:X1=16
```

**Review by relokin — COMMENTED** [review]
Just a small thing I noticed. But I will have another look tomorrow.

File: herd/tests/instructions/AArch64.sve/V23.litmus (hunk 1 above)
Comment (maranget) [reply herd/tests/instructions/AArch64.sve/V23.litmus]: Fixed, thanks @relokin.

**Comment by maranget** [comment]
Hi @relokin, I am currently adding CNT and INC instructions, which look sufficient to write vector length independant code.

File: herd/tests/instructions/AArch64.sve/V23.litmus [hunk 2]
```diff
@@ -0,0 +1,10 @@
+AArch64 V23
+{
+ uint64_t 0:X0;
+ uint64_t 0:X1;
+}
+  P0       ;
+RDVL X0,#0 ;
+RDVL X1,#1 ;
+
+forall 0:X0=0 /\ 0:X1=16
```
[the remark this answers is not in the document]
Comment (maranget) [reply herd/tests/instructions/AArch64.sve/V23.litmus:10]: Indeed, i'll add a comment to the test source

**Comment by maranget** [comment]
Hi @murzinv and @relokin. I'd be glad to have your opinion on the last two commits.

File: herd/tests/instructions/AArch64.sve/V23.litmus (hunk 2 above)
[the remark this answers is not in the document]
Comment (relokin) [reply herd/tests/instructions/AArch64.sve/V23.litmus:10]: When I run this with QEMU="qemu-aarch64 -cpu max,sve-default-vector-length=128" sh run.sh I get.

```
4000000*>0:X0=0; 0:X1=128;
```

Isn't this what herd7 should say as well?

**Comment by maranget** [comment]
Hi @murzinv, I have added your test as `V26`. It works perfectly. Test `V25` was inspired by compiler output...

File: herd/tests/instructions/AArch64.sve/V24.litmus.expected [hunk 3]
```diff
@@ -0,0 +1,11 @@
+Test V24 Required
+States 1
+0:X0=8; 0:X1=8; 0:X2=64; 0:X3=6; 0:X4=0;
```
[the remark this answers is not in the document]
Comment (relokin) [reply herd/tests/instructions/AArch64.sve/V24.litmus.expected:3]: Ah yes you're right last time I used `QEMU="qemu-aarch64 -cpu max,sve128=on` and I assumed that `sve-default-vector-length` is also in bits...I'm going to mark this as resolved.

File: herd/AArch64Sem.ml [hunk 4]
```diff
@@ -3324,6 +3342,22 @@ module Make
...
+           check_sve inst;
+           let sz = MachSize.Quad in
+           let off = scalable_nbytes * k in
+           read_reg_ord_sz sz rn ii
+           >>= M.op1 (Op.AddK off)
+           >>= fun v -> write_reg_sz_dest sz rd v ii
+           >>= nextSet rd
+        | I_CNT_INC_SVE (op,r,pat,k) ->
+           check_sve inst;
+           cnt_inc op r pat k ii >>= nextSet r
```
[the remark this answers is not in the document]
Comment (maranget) [reply herd/AArch64Sem.ml:3360]: The combination of `>>=` and of `nextSet r`  should occur after monad `m` of type `V.v M.t`.  Let `v` be the value returned by m.
The branching effect is the same as `B.Next`. That is, control goes to the next instruction. Additionally, the code that controls instruction symbolic execution (in mem.ml) is informed that the new value of register `r` is `v`.
As an imperative safety condition, `m` should emit a write effect of `v` to register `r`.
Perhaps the most clear example is the combination of `write_reg_dest r .. `  with `>>= nextSet r` as can be found in the implementation of `RBIT`:
```OCaml
 | I_RBIT (v,rd,rn) ->
            let sz = tr_variant v in
            read_reg_ord_sz sz rn ii
            >>= M.op1 (Op.Rbit sz)
            >>= fun v -> write_reg_dest rd v ii
            >>= nextSet rd
```
Here, it looks quite unsurprising that `write_reg_dest rd v ...` emits a write effect to `r`.

File: lib/AArch64Base.ml [hunk 5]
```diff
@@ -1489,6 +1496,12 @@ type 'k kinstruction =
   | I_INDEX_SS of reg * variant * reg * reg
   (* INDEX <Zd>.<T>, #<imm1>, #<imm2> *)
   | I_INDEX_II of reg * 'k * 'k
+  (* RDVL <Xd>,#<imm> *)
+  | I_RDVL of reg * 'k
+  (* ADDVL <Xd>,<Xn>#<imm> *)
```
[the remark this answers is not in the document]
Comment (maranget) [reply lib/AArch64Base.ml:1501]: Done

File: lib/AArch64Base.ml [hunk 6]
```diff
@@ -1489,6 +1496,12 @@ type 'k kinstruction =
...
+  | I_ADDVL of reg * reg * 'k
+(* {CNT,INC}<B|H|SD> <Xd>[,pattern[,MULL #imm]] *)
```
[the remark this answers is not in the document]
Comment (maranget) [reply lib/AArch64Base.ml:1503]: Done, thanks.

**Comment by maranget** [comment]
> > Test V25 was inspired by compiler output...
> 
> Out of curiosity, could you share compiler _**input**_? 😄

I'll look for it.

**Comment by maranget** [comment]
Thanks for your review @murzinv. Let us delay merge as I'd like to simplify history.

**Comment by maranget** [comment]
> > Test V25 was inspired by compiler output...
> 
> Out of curiosity, could you share compiler _**input**_? 😄

Compiler input:
```C
#include <arm_sve.h>
 
void saxpy_c(int32_t *x, int32_t *y) {
        int i;
        for (i=0; i<1023; i++) {
                y[i] = x[i] + y[i];
        }
}
```
Original code is from [this page](https://developer.arm.com/documentation/102699/0100/Optimizing-with-auto-vectorization?lang=en). Compilation command is `gcc -S -march=armv8-a+sve -O2 a.c` on a MacBook M1.
```
% gcc --version
Apple clang version 15.0.0 (clang-1500.1.0.2.5)
Target: arm64-apple-darwin23.2.0
...
```

**Comment by maranget** [comment]
Merged, thanks for teh review @murzinv.

**Pull request merged**